scene/resources/paragraph_layout.cpp

```diff
@@ -0,0 +1,766 @@
+
+	ClassDB::bind_method(D_METHOD("get_offset"), &ClusterInfo::get_offset);
+	ClassDB::bind_method(D_METHOD("get_length"), &ClusterInfo::get_length);
+
+	ClassDB::bind_method(D_METHOD("get_leading_edge"), &ClusterInfo::get_leading_edge);
+	ClassDB::bind_method(D_METHOD("get_tailing_edge"), &ClusterInfo::get_tailing_edge);
+	ClassDB::bind_method(D_METHOD("get_left_edge"), &ClusterInfo::get_left_edge);
+	ClassDB::bind_method(D_METHOD("get_rigth_edge"), &ClusterInfo::get_right_edge);
+
+	ClassDB::bind_method(D_METHOD("get_glyph_count"), &ClusterInfo::get_glyph_count);
+	ClassDB::bind_method(D_METHOD("get_glyph_codepoint", "index"), &ClusterInfo::get_glyph_codepoint);
+	ClassDB::bind_method(D_METHOD("get_glyph_offset", "index"), &ClusterInfo::get_glyph_offset);
+	ClassDB::bind_method(D_METHOD("get_glyph_advance", "index"), &ClusterInfo::get_glyph_advance);
+	ClassDB::bind_method(D_METHOD("get_direction"), &ClusterInfo::get_direction);
+}
+
+int ClusterInfo::get_glyph_count() const {
+
+	return glyphs.size();
+}
+
+uint32_t ClusterInfo::get_glyph_codepoint(int p_index) const {
+
+	if ((p_index < 0) || (p_index >= glyphs.size()))
+		return 0;
+
+	return glyphs[p_index].codepoint;
+}
+
+Point2 ClusterInfo::get_glyph_offset(int p_index) const {
+
+	if ((p_index < 0) || (p_index >= glyphs.size()))
+		return Point2(0, 0);
+
+	return glyphs[p_index].offset;
+}
+
+Point2 ClusterInfo::get_glyph_advance(int p_index) const {
+
+	if ((p_index < 0) || (p_index >= glyphs.size()))
+		return Point2(0, 0);
+
+	return glyphs[p_index].advance;
+}
+
+TextDirection ClusterInfo::get_direction() const {
+
+	return direction;
+}
+
+int ClusterInfo::get_min_char_index() const {
+
+	return min_index;
+}
+
+int ClusterInfo::get_max_char_index() const {
+
+	return max_index;
+}
+
+int ClusterInfo::get_offset() const {
+
+	return offset;
+}
+
+int ClusterInfo::get_length() const {
+
+	return length;
+}
+
+int ClusterInfo::get_leading_edge() const {
+	if (direction == DIR_LTR) {
+		return offset + length;
+	} else {
+		return offset;
+	}
+}
+
+int ClusterInfo::get_tailing_edge() const {
+	if (direction == DIR_LTR) {
+		return offset;
+	} else {
+		return offset + length;
+	}
+}
+
+int ClusterInfo::get_left_edge() const {
+	return offset;
+}
+
+int ClusterInfo::get_right_edge() const {
+	return offset + length;
+}
+
+/*************************************************************************/
+/*  ScriptRun                                                            */
+/*************************************************************************/
+
+UChar32 ScriptRun::paired_chars[] = {
+	0x0028, 0x0029, // ascii paired punctuation
+	0x003c, 0x003e,
+	0x005b, 0x005d,
+	0x007b, 0x007d,
+	0x00ab, 0x00bb, // guillemets
+	0x2018, 0x2019, // general punctuation
+	0x201c, 0x201d,
+	0x2039, 0x203a,
+	0x3008, 0x3009, // chinese paired punctuation
+	0x300a, 0x300b,
+	0x300c, 0x300d,
+	0x300e, 0x300f,
+	0x3010, 0x3011,
+	0x3014, 0x3015,
+	0x3016, 0x3017,
+	0x3018, 0x3019,
+	0x301a, 0x301b
+};
```

Comment:
It might be a good idea to use [`u_getBidiPairedBracket()`](http://icu-project.org/apiref/icu4c/uchar_8h.html#ad88dddebf1a52dc0a74d6e7c2b0dc689) instead of this old, hard-coded list.